### PR TITLE
Support Architecture field

### PR DESCRIPTION
**Pull request by reverson**
## _Read then delete this section_

_- Make sure to use a concise title for the pull-request._

_- Use #patch, #minor or #major in the pull-request title to bump the corresponding version. Otherwise, the patch version
will be bumped. [More details](https://github.com/marketplace/actions/github-tag-bump)_

# TL;DR
Supports architecture field to run on specific CPU architecture

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by reverson** [comment]
@akhurana001

File: go/tasks/config_load_test.go [hunk 1]
```diff
@@ -73,6 +73,44 @@ func TestLoadConfig(t *testing.T) {
...
+		assert.Contains(t, k8sConfig.ArchitectureNodeSelector, "amd64")
+		assert.Equal(t, map[string]string{"x/architecture": "amd64"}, k8sConfig.ArchitectureNodeSelector["amd64"])
+		assert.Contains(t, k8sConfig.ArchitectureTolerations, "amd64")
+		assert.Equal(t, "x/flyte", k8sConfig.ArchitectureTolerations["amd64"][0].Key)
+		assert.Equal(t, "amd64", k8sConfig.ArchitectureTolerations["amd64"][0].Value)
+		assert.NotNil(t, k8sConfig.DefaultPodSecurityContext)
+		assert.NotNil(t, k8sConfig.DefaultPodSecurityContext.FSGroup)
+		assert.Equal(t, *k8sConfig.DefaultPodSecurityContext.FSGroup, int64(2000))
+		assert.NotNil(t, k8sConfig.DefaultPodSecurityContext.RunAsGroup)
+		assert.Equal(t, *k8sConfig.DefaultPodSecurityContext.RunAsGroup, int64(3000))
```
Comment (akhurana001) [inline go/tasks/config_load_test.go:85]: looks like these are unrelated to this change ?

Comment (reverson) [reply go/tasks/config_load_test.go]: Yep, didn't see that one

File: go/tasks/pluginmachinery/core/exec_metadata.go [hunk 2]
```diff
@@ -35,4 +35,6 @@ type TaskExecutionMetadata interface {
 	GetK8sServiceAccount() string
 	GetSecurityContext() core.SecurityContext
 	IsInterruptible() bool
+	GetArchitecture() core.Container_Architecture
+	GetPlatformResources() *v1.ResourceRequirements
```
[the remark this answers is not in the document]
Comment (reverson) [reply go/tasks/pluginmachinery/core/exec_metadata.go]: Yep, the rebase didn't work properly, fixing these up.

File: go/tasks/plugins/k8s/spark/spark.go [hunk 3]
```diff
@@ -188,6 +188,14 @@ func (sparkResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsCo
 	}
 	executorSpec.Memory = strPtr(sparkConfig["spark.executor.memory"])
 
+	// Add Architecture Tolerations/NodeSelector to all pods
+	if taskTemplate.GetContainer().GetArchitecture() != core.Container_UNKNOWN {
+		driverSpec.SparkPodSpec.Tolerations = config.GetK8sPluginConfig().ArchitectureTolerations[strings.ToLower(taskTemplate.GetContainer().GetArchitecture().String())]
```
Comment (akhurana001) [inline go/tasks/plugins/k8s/spark/spark.go:193]: won't this and interruptible overwrite each other ?

File: go/tasks/plugins/k8s/spark/spark.go [hunk 4]
```diff
@@ -217,11 +217,21 @@ func (sparkResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsCo
 		j.Spec.MainClass = &sparkJob.MainClass
 	}
 
+	// Add Architecture Tolerations/NodeSelector to all pods
+	if taskTemplate.GetContainer().GetArchitecture() != core.Container_UNKNOWN {
+		j.Spec.Driver.NodeSelector = config.GetK8sPluginConfig().ArchitectureNodeSelector[strings.ToLower(taskTemplate.GetContainer().GetArchitecture().String())]
+		j.Spec.Driver.Tolerations = config.GetK8sPluginConfig().ArchitectureTolerations[strings.ToLower(taskTemplate.GetContainer().GetArchitecture().String())]
+		j.Spec.Executor.NodeSelector = config.GetK8sPluginConfig().ArchitectureNodeSelector[strings.ToLower(taskTemplate.GetContainer().GetArchitecture().String())]
+		j.Spec.Executor.Tolerations = config.GetK8sPluginConfig().ArchitectureTolerations[strings.ToLower(taskTemplate.GetContainer().GetArchitecture().String())]
+		j.Spec.SparkConf["spark.kubernetes.node.selector.beta.kubernetes.io/arch"] = "arm64"
```
Comment (akhurana001) [inline go/tasks/plugins/k8s/spark/spark.go:226]: this is supported by core-spark afaik. The operator is supported to convert the Spec entries above to eventually spark conf like this but looks like there is a bug there.
